indra/newview/llfloaterimnearbychatlistener.cpp

```diff
@@ -49,6 +50,34 @@ LLFloaterIMNearbyChatListener::LLFloaterIMNearbyChatListener()
         "[\"channel\"] chat channel number [default = 0]\n"
         "[\"type\"] chat type \"whisper\", \"normal\", \"shout\" [default = \"normal\"]",
         &LLFloaterIMNearbyChatListener::sendChat);
+
+    add("listen",
+        "Start listening to the Nearby chat, chat messages will be resent to the script event pump",
+        &LLFloaterIMNearbyChatListener::listenChat,
+        llsd::map("reply", LLSD()));
+
+    add("stopListening",
+        "Stop listening to the Nearby chat",
+        &LLFloaterIMNearbyChatListener::stopListeningChat);
+ 
+    mOutConnection = LLEventPumps::instance().obtain("LLNearbyChat").listen("LLFloaterIMNearbyChatListener", [this](const LLSD &data)
+    {
+        std::map<std::string, std::string> reply_pumps = mReplyPumps;
+        std::map<std::string, std::string> scripts = LLLUAmanager::getScriptNames();
+        for (auto &it : reply_pumps)
+        {
+            //check if listener script is still running
+            if (scripts.find(it.first) != scripts.end())
+            {
+                LLEventPumps::instance().obtain(it.second).post(data);
+            }
+            else
+            {
+                mReplyPumps.erase(it.first);
+            }
+        }
+        return false;
+    });
```

Comment:
`LLLeapListener` (accessed via the script's "command pump") already supports this functionality via the "listen" and "stoplistening" operations. Since "listen" stores the new connection in the `LLLeapListener` instance itself, and since each `LuaState` has its own `LuaListener`, and each `LuaListener` its own `LLLeapListener` instance, the connection is guaranteed to disconnect on script termination (`~LuaState()`).

Reply:
I see, I totally missed that and tried to reinvent the wheel.

Reply:
Part of the problem here is that I have never written any documentation about the available `LLEventAPI` listeners and the operations they support, so discoverability is low. In my defense, I always expected the suite of listeners and their operations to keep growing (and to diverge across branches and viewer forks), so that the only reliable information is really the source code. That's why every listener and every operation is defined with queryable metadata.